### PR TITLE
Rasterize shape during geotilegrid computation

### PR DESCRIPTION
**Pull request by talevy**
This PR mainly modifies the existing GeoTileGridTiler to rasterize
the GeometryTree instead of iterating through all the tiles found in the 
bounding box of the shape.

This PR also fixes a bug where containsFully was not being calculated correctly

relates #37206.

### CODE REVIEW

**Comment by iverase** [comment]
Thanks @talevy! One thing I was expecting is just one method that returns the relationship between the docValue and the Extent, something like:

```
Relation relate(Extent extent);
`` 

Instead you are using two methods, one for `within` and one for `intersects`, is there any technical reason for doing that?

**Comment by talevy** [comment]
I started with this route just to make sure the logic was right since conflating both `intersects` and `within` in the same method would have made things rather complicated.

two technical reasons for splitting:

- simpler code
- `intersects` can exit early while `within` requires a full tree traversal.

Main reason to join the two

- The `GeoTileGridTiler#setValuesForCell` code first calls `within` and then falls back to `intersects`. Combining both into a `relate` would make this go faster since right now there is unnecessary duplicate de-serialization and traversal.

Since the only callers of these methods are the tilers, then maybe combining makes sense. I will try it out and see how it affects performance 👍

**Comment by talevy** [comment]
One example of where it may be nice to have separate within/intersects is with the planned hexgrid aggregation where `within` is likely never to be asked: 

example photo: http://1fykyq3mdn5r21tpna3wkdyi-wpengine.netdna-ssl.com/wp-content/uploads/2018/06/image6-1.png

**Comment by iverase** [comment]
This approach will not work for hex grid as a parent cell must contain all the children cells. You can have a shape that is disjoint with the parent cell but intersects a children cell.

**Comment by talevy** [comment]
right, there is no concept of "within" in that case. I don't bring it up in the context of rasterization, but just in the context of keeping intersect and within separate.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java [hunk 1]
```diff
@@ -121,24 +123,44 @@ public long getBoundingTileCount(MultiGeoValues.GeoValue geoValue, int precision
...
+                return valuesIndex;
             }
 
-            return idx;
+            if (GeoRelation.QUERY_INSIDE == relation) {
+                return setValuesForFullyContainedTile(tile, values, valuesIndex, precision);
+            }
+            if (GeoRelation.QUERY_CROSSES == relation) {
+                int[][] subTiles = GeoTileUtils.getSubTiles(tile[0], tile[1], tile[2]);
+                for (int[] subTile : subTiles) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: I think we do not need to call that method and create the integer arrays, something like:

```
for (int I = 0; i < 2; i++) {
   for (int j = 0; j < 2; j++) {
       int nextX = 2 * x + i;
       int nextY = 2 * y + j;
       Rectangle r = GeoTileUtils.toBoundingBox(nextX, nextY, zoom + 1);
       valuesIndex = setValuesForCell(subTile, values, valuesIndex, precision, geoValue);
  }
}
```

WDYT?

Comment (talevy) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: yeah! I am still getting used to the "do not create objects" way of writing code. I'll update. thanks!

File: server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java [hunk 2]
```diff
@@ -48,25 +51,27 @@ public Extent getExtent() throws IOException {
     /**
      * Returns true if the rectangle query and the edge tree's shape overlap
      */
-    public boolean intersects(Extent extent) throws IOException {
-        if (hasArea) {
-            return containsBottomLeft(extent) || crosses(extent);
-        } else {
-            return crosses(extent);
+    @Override
+    public GeoRelation relate(Extent extent) throws IOException {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java]: I think we can move the check of the extent here. Not needed for this PR but I think it is worthy to note.

**Review by iverase — APPROVED** [review]
I added a note where we can improve slightly things but not needed for this PR. I think this is an massive improvement so lgtm.

File: server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java (hunk 2 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/EdgeTreeReader.java]: I agree. consolidating to one relating method reduces the places that there is an extent check. I will follow-up

**Pull request merged**